File: algorithm/leetcode971.flip-binary-tree-to-match-preorder-traversal.cpp
class Solution {
	int i = 0;
	bool can = true;
	vector<int> ans;
public:
    vector<int> flipMatchVoyage(TreeNode* root, vector<int>& voyage) {
    	dfs(root, voyage);
    	if(!can) return {-1};
    	return ans;
    }

    void dfs(TreeNode* root, vector<int>& voyage)
    {
    	if(!can || !root)
    		return;
    	if(root->val == voyage[i])
    	{
    		i++;
    		if(root->left && root->left->val == voyage[i])
    		{
				dfs(root->left, voyage);
				dfs(root->right, voyage);
    		}
    		else if(root->right && root->right->val == voyage[i])
    		{
                if(root->left)
    			    ans.push_back(root->val);
    			dfs(root->right, voyage);
    			dfs(root->left, voyage);
    		}
    		else if(root->left || root->right)
    			can = false;
    	}
    	else
    		can = false;
    }
};

File: algorithm/Python3_version/leetcode971.flip-binary-tree-to-match-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.i = 0
        self.can = True
        self.ans = []
    def flipMatchVoyage(self, root: TreeNode, voyage: List[int]) -> List[int]:
        def dfs(root, voyage):
            if not self.can or not root:
                return
            if root.val == voyage[self.i]:
                self.i += 1
                if root.left and root.left.val == voyage[self.i]:
                    dfs(root.left,voyage)
                    dfs(root.right, voyage)
                elif root.right and root.right.val == voyage[self.i]:
                    if root.left:
                        self.ans.append(root.val)
                    dfs(root.right, voyage)
                    dfs(root.left, voyage)
                elif root.left or root.right:
                    self.can = False
            else:
                self.can = False
        dfs(root, voyage)
        if not self.can:
            return [-1]
        return self.ans